ampd/src/stellar/rpc_client.rs

```diff
@@ -0,0 +1,120 @@
+use std::collections::{HashMap, HashSet};
+use std::str::FromStr;
+
+use error_stack::{report, ResultExt};
+use futures::future::join_all;
+use stellar_rpc_client::GetTransactionResponse;
+use stellar_xdr::curr::{ContractEvent, Hash, TransactionMeta, VecM};
+use thiserror::Error;
+use tracing::warn;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to create client")]
+    Client,
+    #[error("invalid tx hash")]
+    TxHash,
+}
+
+/// TxResponse parses XDR encoded TransactionMeta to ContractEvent type, and only contains necessary fields for verification
+#[derive(Debug)]
+pub struct TxResponse {
+    pub transaction_hash: String,
+    pub successful: bool,
+    pub contract_events: VecM<ContractEvent>,
+}
+
+const STATUS_SUCCESS: &str = "SUCCESS";
+
+impl From<(Hash, GetTransactionResponse)> for TxResponse {
+    fn from((transaction_hash, response): (Hash, GetTransactionResponse)) -> Self {
+        let contract_events = response
+            .result_meta
+            .and_then(|meta| match meta {
+                TransactionMeta::V3(data) => data.soroban_meta.map(|meta| meta.events),
+                _ => None,
+            })
+            .unwrap_or_else(VecM::default);
+
+        Self {
+            transaction_hash: transaction_hash.to_string(),
+            successful: response.status == STATUS_SUCCESS,
+            contract_events,
+        }
+    }
+}
+
+impl TxResponse {
+    pub fn has_failed(&self) -> bool {
+        !self.successful
+    }
+
+    pub fn event(&self, index: u64) -> Option<&ContractEvent> {
+        let log_index = usize::try_from(index).ok()?;
+        self.contract_events.get(log_index)
+    }
+
+    pub fn tx_hash(&self) -> String {
+        self.transaction_hash.clone()
+    }
+}
+
+#[cfg_attr(test, faux::create)]
+pub struct Client(stellar_rpc_client::Client);
+
+#[cfg_attr(test, faux::methods)]
+impl Client {
+    pub fn new(url: String) -> error_stack::Result<Self, Error> {
+        Ok(Self(
+            stellar_rpc_client::Client::new(url.as_str())
+                .map_err(|err_str| report!(Error::Client).attach_printable(err_str))?,
+        ))
+    }
+
+    pub async fn transaction_responses(
+        &self,
+        tx_hashes: HashSet<String>,
+    ) -> error_stack::Result<HashMap<String, TxResponse>, Error> {
+        let tx_hashes: Vec<_> = tx_hashes
+            .into_iter()
+            .map(|tx_hash| Hash::from_str(tx_hash.as_str()).change_context(Error::TxHash))
+            .collect::<std::result::Result<Vec<_>, _>>()?;
+
+        let responses = join_all(
+            tx_hashes
+                .iter()
+                .map(|tx_hash| self.0.get_transaction(tx_hash)),
```

Comment:
Looks like we don't support doing a batched JSON RPC query in ampd? Seems to be the same as EVM. Is there a task to add this later?

Reply:
The crate doesn’t support batch yet; we can update once they add that support.